File: app/src/main/java/de/davis/passwordmanager/ui/sync/ImportActivity.java
package de.davis.passwordmanager.ui.sync;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;

import de.davis.passwordmanager.backup.DataBackup;
import de.davis.passwordmanager.backup.keygo.KeyGoBackup;
import de.davis.passwordmanager.databinding.ActivityImportBinding;
import de.davis.passwordmanager.ui.MainActivity;
import de.davis.passwordmanager.ui.login.LoginActivity;

public class ImportActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ActivityImportBinding binding = ActivityImportBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        ActivityResultLauncher<Intent> auth = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(), result -> {
            if(result.getResultCode() != RESULT_OK)
                return;

            Intent intent = getIntent();
            if(intent == null || intent.getAction() == null)
                return;

            if (!intent.getAction().equals(Intent.ACTION_VIEW))
                return;

            Uri fileUri = intent.getData();
            if(fileUri == null)
                return;

            KeyGoBackup backup = new KeyGoBackup(this);
            backup.execute(DataBackup.TYPE_IMPORT, fileUri, r -> {
                startActivity(new Intent(this, MainActivity.class));
            });
        });

        auth.launch(LoginActivity.getIntentForAuthentication(this));
        binding.button.setOnClickListener(v -> auth.launch(LoginActivity.getIntentForAuthentication(this)));
    }
}

File: app/src/main/java/de/davis/passwordmanager/ui/sync/ImportActivity.kt
package de.davis.passwordmanager.ui.sync

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import de.davis.passwordmanager.backup.BackupOperation
import de.davis.passwordmanager.backup.BackupResult
import de.davis.passwordmanager.backup.impl.AndroidBackupListener
import de.davis.passwordmanager.backup.impl.AndroidPasswordProvider
import de.davis.passwordmanager.backup.impl.KdbxBackup
import de.davis.passwordmanager.backup.impl.UriStreamProvider
import de.davis.passwordmanager.databinding.ActivityImportBinding
import de.davis.passwordmanager.ui.MainActivity
import de.davis.passwordmanager.ui.auth.AuthenticationRequest
import de.davis.passwordmanager.ui.auth.createRequestAuthenticationIntent
import kotlinx.coroutines.launch

class ImportActivity : AppCompatActivity() {

    private val kdbxBackup = KdbxBackup(
        AndroidPasswordProvider(this),
        object : AndroidBackupListener(this@ImportActivity) {

            override fun onSuccess(
                backupOperation: BackupOperation,
                backupResult: BackupResult
            ) {
                super.onSuccess(backupOperation, backupResult)
                startActivity(
                    Intent(
                        this@ImportActivity,
                        MainActivity::class.java
                    )
                )
                finish()
            }
        })

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding = ActivityImportBinding.inflate(
            layoutInflater
        )
        setContentView(binding.root)

        val auth =
            registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result: ActivityResult ->
                if (result.resultCode != RESULT_OK) return@registerForActivityResult
                if (intent.action == null) return@registerForActivityResult
                if (intent.action != Intent.ACTION_VIEW) return@registerForActivityResult
                val fileUri = intent.data ?: return@registerForActivityResult

                lifecycleScope.launch {
                    kdbxBackup.execute(
                        BackupOperation.IMPORT,
                        UriStreamProvider(fileUri, contentResolver)
                    )
                }
            }
        val authIntent = createRequestAuthenticationIntent(AuthenticationRequest.JUST_AUTHENTICATE)
        auth.run {
            launch(authIntent)
            binding.button.setOnClickListener { _: View? -> launch(authIntent) }
        }
    }
}